lib/SILOptimizer/Mandatory/ConstantPropagation.cpp

```diff
@@ -985,10 +1032,297 @@ constantFoldStringConcatenation(ApplyInst *AI,
+    }
+    // This is something we cannot handle.
+    return nullptr;
+  }
+  // Bail if there was no assignment.
+  if (!SI)
+    return nullptr;
+  // Bail if it was not an assignment of an integer literal.
+  auto StructI = dyn_cast<StructInst>(SI->getSrc());
+  if (!StructI)
+    return nullptr;
+  return dyn_cast<IntegerLiteralInst>(StructI->getOperand(0));
+}
+
+static IntegerLiteralExpr *
+getIntegerLiteralExprFromConstructorCall(ApplyInst *AI) {
+  auto Loc = AI->getLoc();
+  const ApplyExpr *Apply = Loc.getAsASTNode<ApplyExpr>();
+  const Expr *ValueExpr = Apply->getArg();
+  if (auto *Paren = dyn_cast<ParenExpr>(ValueExpr))
+    ValueExpr = Paren->getSubExpr();
+  auto *CallExp = dyn_cast<CallExpr>(ValueExpr);
+  if (CallExp) {
+    auto Callee = CallExp->getFn();
+    if (isa<ConstructorRefCallExpr>(Callee) &&
+        CallExp->getNumArguments() == 1) {
+      // Check if takes an integer literal as parameter.
+      auto Arg = CallExp->getArg();
+      auto ArgLabels = CallExp->getArgumentLabels();
+      if (ArgLabels.size() == 1 &&
+          ArgLabels[0].str() == "_builtinIntegerLiteral") {
+        if (auto TupleExp = dyn_cast<TupleExpr>(Arg)) {
+          return dyn_cast<IntegerLiteralExpr>(TupleExp->getElement(0));
+        }
+      }
+    }
+  }
+  return nullptr;
+}
+
+/// Special handling of the generic initializers provided by
+/// UnsignedInteger and SignedInteger protocols when they are
+/// invoked for fixed size builtin integer types and constant
+/// arguments.
+///
+/// Check a call of an initializer for one of the builtin integer types.
+/// Try to determine if the initializer takes an integer constant
+/// as an argument and if this constant overflows/underflows the values
+/// representable by this type.
+///
+/// The initializer in question is one of:
+/// UnsignedInteger.init<T : BinaryInteger>(_ source: T), or
+/// SignedInteger.init<T : BinaryInteger>(_ source: T)
+///
+/// A typical invocation of such an initializer looks in SIL like this:
+///
+///   // function_ref UnsignedInteger<A>.init<A>(_:)
+///   %init = function_ref @init_name: $@convention(method)
+///     <τ_0_0 where τ_0_0 : FixedWidthInteger, τ_0_0 : UnsignedInteger>
+///     <τ_1_0 where τ_1_0 : BinaryInteger>
+///     (@in τ_1_0, @thick τ_0_0.Type) -> @out τ_0_0
+///   %int_lit = integer_literal $Builtin.Int64, 400
+///   %int_val = struct $Int (%int_lit : $Builtin.Int64)
+///   %result = alloc_stack $UInt8
+///   %tmp = alloc_stack $Int
+///   %ty = metatype $@thick UInt8.Type
+///   store %int_val to %tmp : $*Int
+///   apply %init<UInt8, Int>(%result, %tmp, %ty) : ...
+///   dealloc_stack %tmp : $*Int
+bool ConstantPropagation::checkIntegerInitializer(
+    ApplyInst *AI, llvm::SetVector<SILInstruction *> &WorkList) {
+  auto &Ctx = AI->getModule().getASTContext();
+
+  assert(AI->getSubstitutions().size() == 2);
+
+
+  // First check source and destination types.
+  auto GetStdlibIntTypeInfo = [](CanType Ty, unsigned &TySize,
```

Comment:
Not really. I could easily move it out of the function into a separate static function.